sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc

```diff
@@ -0,0 +1,379 @@
+= sycl_ext_oneapi_complex_marray
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2022-2022 Codeplay Ltd.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 5 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+The complex marray extension builds on the `sycl::ext::oneapi::complex` class,
+therefore this extension is dependent on
+link:sycl_ext_oneapi_complex.asciidoc[sycl_ext_oneapi_complex].
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+[NOTE]
+====
+This extension is not currently implemented in {dpcpp}.
+====
+
+
+== Overview
+
+{dpcpp} has support for `sycl::ext::oneapi::complex` this allows for the
+addition of new complex math features. This proposes the instantiation of
+`marray` to add support for complex numbers to be stored in arrays. The
+proposal includes overloading the existing math functions to support complex
+marrays and adding new free functions to simplify accessing, setting marray
+values, and construct complex marrays.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_COMPLEX_MARRAY` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== Marray Complex Class Specialization
+
+The `marray` class is extended for the `sycl::ext::oneapi::complex` class.
```

Comment:
Why is that needed? SYCL 2022 says "The element type parameter, DataT, must be a numeric type as it is defined by C++ standard.". Therefore marray already supports complex. I think everything up line 106 isn't needed.

Reply:
So I am not sure that `complex` is a numeric type. So far as I understand (please correct me if I am wrong), a type is numeric if it has `std::numeric_limits` defined for it. `std::complex` does not appear to have a numeric limits specialization. While `std::numeric_limits` could be specialized for complex, a lot of the constants and member functions don't make sense. For example, what is the complex equivalent of infinity, is it `(inf, inf)` does that mean `(inf, 0)` is not infinity although its magnitude is infinity?

I think the safer and simpler way of viewing `complex` is as two numeric types, but it itself is not a numeric type.

Reply:
Here is the cppreference definition of `NumericType`:

https://en.cppreference.com/w/cpp/named_req/NumericType

It does not say that `std::numeric_limits` must be defined.

My feedback on these lines through 106:

* I agree with @rolandschulz that the line "The `marray` class is extended for the `sycl::ext::oneapi::complex` class." is unneeded.  I do not think we are extending that class.  In fact, the next sentence says that the interface is not changed, so these two statements seems to be contradictory.

* The statement that "The `marray` complex instantiation is trivially copyable" seems useful.  I don't think we currently have anything in the SYCL 2020 spec that guarantees this, though there is an open Khronos issue about it (https://gitlab.khronos.org/sycl/Specification/-/issues/564).

* The statement "The `marray` definition used within this proposal assumes that any operator the `marray` class defines is only implemented if the marray's value type also implements the operator" also seems useful.  This is also missing from SYCL 2020, but we have a Khronos issue to add wording like this (https://gitlab.khronos.org/sycl/Specification/-/issues/559).

Reply:
> * I agree with @rolandschulz that the line "The `marray` class is extended for the `sycl::ext::oneapi::complex` class." is unneeded.  I do not think we are extending that class.  In fact, the next sentence says that the interface is not changed, so these two statements seems to be contradictory.

This makes sense to me. I have removed this line.